packages/assets-controllers/src/MultichainTokensRatesController/MultichainTokensRatesController.ts

```diff
@@ -0,0 +1,391 @@
+   *
+   * @returns True if the keyring is unlocked; otherwise, false.
+   */
+  get isActive(): boolean {
+    return this.#isUnlocked;
+  }
+
+  /**
+   * Checks if an account is a non-EVM account with a Snap.
+   *
+   * @param account - The account to check.
+   * @returns True if the account is non-EVM and has Snap metadata; otherwise, false.
+   */
+  #isNonEvmAccount(account: InternalAccount): boolean {
+    return (
+      !isEvmAccountType(account.type) && account.metadata.snap !== undefined
+    );
+  }
+
+  /**
+   * Retrieves all multichain accounts from the AccountsController.
+   *
+   * @returns An array of internal accounts.
+   */
+  #listMultichainAccounts(): InternalAccount[] {
+    return this.messagingSystem.call(
+      'AccountsController:listMultichainAccounts',
+    );
+  }
+
+  /**
+   * Filters and returns non-EVM accounts that should have balances.
+   *
+   * @returns An array of non-EVM internal accounts.
+   */
+  #listAccounts(): InternalAccount[] {
+    const accounts = this.#listMultichainAccounts();
+    return accounts.filter((account) => this.#isNonEvmAccount(account));
+  }
+
+  /**
+   * Retrieves a non-EVM account by its ID.
+   *
+   * @param accountId - The account ID.
+   * @returns The corresponding internal account, or undefined if not found.
+   */
+  #getAccount(accountId: string): InternalAccount | undefined {
+    return this.#listAccounts().find((account) => account.id === accountId);
+  }
+
+  /**
+   * Handles the removal of an account by deleting its conversion rates.
+   *
+   * @param accountId - The ID of the removed account.
+   */
+  async #handleOnAccountRemoved(accountId: string): Promise<void> {
+    if (accountId in this.state.conversionRates) {
+      this.update((state: Draft<MultichainTokensRatesControllerState>) => {
+        delete state.conversionRates[accountId];
+      });
+    }
+  }
+
+  /**
+   * Updates the token conversion rates for a given account.
+   *
+   * This method acquires a mutex lock to ensure thread safety.
+   *
+   * @param accountId - The account ID for which to update rates.
+   * @returns A promise that resolves when the rates have been updated.
+   */
+  async updateTokensRates(accountId: string): Promise<void> {
+    const releaseLock = await this.#mutex.acquire();
+
+    return (async () => {
+      if (!this.isActive) {
+        return;
+      }
+
+      const account = this.#getAccount(accountId);
+
+      if (!account?.metadata.snap) {
+        return;
+      }
+
+      // TODO: Remove this condition when additional Snap support is implemented.
+      if (account.type !== 'solana:data-account') {
+        return;
+      }
+
+      // Retrieve assets from the assets controller.
+      const assets = await this.#getAssetsList(
+        accountId,
+        account.metadata.snap.id as SnapId,
+      );
+
+      const conversions = assets.map((asset) => ({
+        from: asset,
+        to: SWIFT_ISO4217_USD,
```

Comment:
Why are we hardcoding USD ? This should rely on the user's preferred currency to set the target

Reply:
I agree, but the price API doesn't support every currency that the user can choose. As mentioned in [this thread](https://consensys.slack.com/archives/C07M5L4U51U/p1738328615066659)

Reply:
we are not hardcoding USD any mode , instead we're trying to get price on native token ( SOL for solana, BTC for Bitcoin, ETH for Ethereum ... etc )

Reply:
This is something that the snap itself will need to deal with for a couple of reasons:

1. We won't always know what API a snap will be using to provide rates
2. This means we will have to add more logic on the extension side to convert from USD to the user's preferred currency.

Reply:
yes this will add an additional computation on the UI side indeed, but actually the snap is relying on price api , and price api is not supporting the list on currencies below that the user can choose.
we have actually the same computation on the UI side for EVM as well.
<img width="1182" alt="Screenshot 2025-01-31 at 12 58 53" src="https://github.com/user-attachments/assets/3c7534de-3b1e-4c21-9566-c3440fa81942" />

Reply:
we can implement this logic on the snap side ( which will be better ) , but for now the version i'm using is making sample call to price api

Reply:
I agree with @GuillaumeRx on this, it makes more sense for the controller to request the actual fiat conversion we want and have the Snap deal with providing it even if the API doesn't support it. The Snap can support it by doing the additional computation, we don't need to bring that complexity into the controller IMO.

Reply:
IMO the controller should just be here to interface between MetaMask and the snap. We shouldn't fix issues in it. This creates an unwanted authority between the client and the snap.

Reply:
@GuillaumeRx @FrederikBolding  alright, who can help me to implement this on the snap side ? i'll ask @aganglada if he can do it

Reply:
@salimtb agree with the above, in order to request a conversion rate, you pass your conversions, we deal with them. No need to worry about the snap. 
The snap right now calls platform price api, which does not suport several conversions, but they will fix them.